File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        pre = [0] * len(nums)
        pre[0] = nums[0]
        post = [0] * len(nums)
        post[len(post) - 1] = nums[len(nums) - 1]

        
        for i in range(1, len(nums) - 1):
            pre[i] = pre[i - 1] * nums[i]
        for i in range(len(nums) - 2, 0, -1):
            post[i] = post[i + 1] * nums[i]

        res = []
        res.append(post[1])
        for i in range(1, len(nums) - 1):
            res.append(post[i + 1] * pre[i - 1])
        res.append(pre[len(nums) - 2])
        return res

File: Data Structures & Algorithms/three-integer-sum/submission-0.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        for (int i = 0; i < (int)nums.size() - 2; i++) {
            if (i > 0 && nums[i-1] == nums[i]) {
                continue;
            }
            int l = i + 1;
            int r = nums.size() - 1;
            int target = -1*nums[i];
            while (r > l) {
                if (nums[l] + nums[r] == target) {
                    res.push_back({nums[i], nums[l], nums[r]});
                    while (l < r && nums[l] == nums[l+1]) l++;
                    while (l < r && nums[r] == nums[r-1]) r--;
                    l++;
                    r--;
                } else if (nums[l] + nums[r] > target) {
                    r--;
                } else {
                    l++;
                }
            }

        }

        return res;
    }
};
